### PR TITLE
Simplify weak types in promotion table

### PR DESCRIPTION
**Pull request by jakevdp**
I realized that the custom class is not necessary.

A side-effect of this change is that `jnp.promote_types` will now treat `int`, `float`, and `complex` as weakly-typed, which is consistent with how they are handled internally; e.g.
```python
>>> jnp.promote_types(jnp.float16, float)                                                                                                                                
dtype('float16')
```
This is a breaking change, in that this previously returned `dtype('float64')`, but I think it's the right API to have here.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
LGTM, if this is a breaking change, should we do a global check for internal failures before merging?

**Comment by jakevdp** [comment]
I made a minor tweak that causes `promote_types` to be backward compatible; internal global check passes.

**Pull request merged**
